### PR TITLE
Dynamic prometheus

### LINKED ISSUE #1125: Dynamic Proxying Support For Prometheus Plugin
The multi-server proxy config approach for prometheus is great, but in a very large company with > 100 prometheus server, maintaining this centralized proxy config in backstage becomes a bottleneck. Many teams aren't quite sure how to update it, and it just adds more onboarding steps.

I'd like to be able to control the backend proxy behavior myself using my own proxy middleware functionality. This would be an optional advanced setup.

## Possible Implementation

On the plugin side, the change is very simple. All we would end up doing is update the requests to the proxy backend to additionally pass the `prometheus.io/service-name` annotation as a request header.

```tsx
const response = await fetch(
  `${apiUrl}/query_range?query=${query}&start=${start}&end=${end}&step=${step}`, {
    headers: {
      'x-prometheus-service-name': serviceName || '',
    },
  }
);
```

The backstage maintainer team then can come up with their own conventions on how to dynamically turn that service-name into a URL to proxy to. Maybe all teams have a standard prometheus setup deployed as a subdomain to a common domain. You can write your own proxy routing to achieve this

Step 1: Update app-config to use a special path if it can't find the `prometheus.io/service-name` config
**app-config.yaml**
```yaml
prometheus:
  proxyPath: '/dynamic-prometheus'
```

Step 2: Hijack this path by writing your own proxy middleware extension
**src/plugins/proxy.ts**
```ts
export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  const proxyRouter = await createRouter({
    logger: env.logger,
    config: env.config,
    discovery: env.discovery,
  });
  const externalUrl = await env.discovery.getExternalBaseUrl('proxy');
  const { pathname: pathPrefix } = new URL(externalUrl);
  proxyRouter.use(
    '/dynamic-prometheus',
    createProxyMiddleware({
      logProvider: () => env.logger,
      logLevel: 'debug',
      changeOrigin: true,
      pathRewrite: {
        [`^${pathPrefix}/dynamic-prometheus/?`]: '/',
      },
      router: req => {
        const prometheusServiceName = req.headers['x-prometheus-service-name'];
        // some code that does something with the prometheusServiceName
        // this could be as simple as adding it as a subdomain to a standard baseURL
        // Likely implementers may want to do some basic validation.
        return `https://${prometheusServiceName}.company.com`;
      },
    }),
  );

  return proxyRouter;
}

```

### PR DESCRIPTION
**Pull request by bforbis**
Closes #1125

#### :heavy_check_mark: Checklist

- [ ] Added tests for new functionality and regression tests for bug fixes
- [x] Added changeset (run `yarn changeset` in the root)
- [ ] Screenshots of before and after attached (for UI changes)
- [x] Added or updated documentation (if applicable)

### CODE REVIEW

**Review by Xantier — APPROVED** [review]
Nice! Thanks

File: plugins/frontend/backstage-plugin-prometheus/README.md [hunk 1]
```diff
@@ -191,14 +186,63 @@ prometheus:
...
+
 ## Using callback with `EntityPrometheusAlertCard`
 
 You can add callbacks that will be executed when the user clicks on a row in the table of the `EntityPrometheusAlertCard` component. This callback is optional, and the rows become clickable only when the callback is supplied. The callback function is also provided with data of type JSON. It contains the information from the row of the table.
 The component with callback can look like this:
 
 ```typescript
 <EntityPrometheusAlertCard onRowClick={callbackFunction} />
-```
+````
```
Comment (bforbis) [inline plugins/frontend/backstage-plugin-prometheus/README.md:245]: Looks like I screwed up the code fence. Will fix this.

File: plugins/frontend/backstage-plugin-prometheus/README.md [hunk 2]
```diff
@@ -191,6 +186,60 @@ prometheus:
...
+
+```yaml
+prometheus:
+  proxyPath: '/dynamic-prometheus'
+```
+
+Step 2: Hijack this path by writing your own proxy middleware extension.
+**packages/backend/src/plugins/proxy.ts**
+
+````diff
```
Comment (Xantier) [inline plugins/frontend/backstage-plugin-prometheus/README.md:206]:
```suggestion
```diff
```

**Comment by Xantier** [comment]
> CI run seems to not have downloaded Cypress correctly.

Yup, GitHub does tht sometimes. I'll kick it off again

**Review by Xantier — APPROVED** [review]

**Comment by Xantier** [comment]
Was thinking of releasing it with upgrades to 1.18, but I just shipped this now as a standalone shipment

**Pull request merged**